src/Workleap.Extensions.Mongo.Ephemeral/ReusableMongoRunnerProvider.cs

```diff
@@ -1,38 +1,90 @@
 using System.Collections.Concurrent;
+using System.Globalization;
+using EphemeralMongo;
 
 namespace Workleap.Extensions.Mongo.Ephemeral;
 
+#pragma warning disable EMEX0001 // MongoRunnerPool is marked as experimental
+
 internal sealed class ReusableMongoRunnerProvider : IDisposable
 {
-    private static readonly ConcurrentDictionary<string, Lazy<ReusableMongoRunner>> LazyRunners = new(StringComparer.Ordinal);
+    private static readonly ConcurrentDictionary<string, Lazy<MongoRunnerPool>> LazyRunnerPools = new(StringComparer.Ordinal);
 
-    private readonly Guid _renterId;
+    private readonly Dictionary<string, List<IMongoRunner>> _rentedRunnersByClientName = [];
 
-    public ReusableMongoRunnerProvider()
+    public IMongoRunner GetRunner(string clientName)
     {
-        this._renterId = Guid.NewGuid();
+        var runner = LazyRunnerPools.GetOrAdd(clientName, CreateLazyMongoRunnerPool).Value.Rent();
+
+        lock (this._rentedRunnersByClientName)
+        {
+            if (this._rentedRunnersByClientName.TryGetValue(clientName, out var rentedRunners))
```

Comment:
I don't think perf is important here, but otherwise you can use `CollectionsMarshal.GetValueRefOrAddDefault`. Or maybe it's easier to use a `ConcurrentDictionary` and remove the `lock`

````c#
public static TValue GetOrAdd<TKey, TValue>(this Dictionary<TKey, TValue> dict, TKey key, TValue value)
    where TKey : notnull
{
    ref var dictionaryValue = ref CollectionsMarshal.GetValueRefOrAddDefault(dict, key, out var exists);
    if (exists)
        return dictionaryValue;

    dictionaryValue = value;
    return value;
}
````

Reply:
Changed to use concurrent dict, not sure why I didn't went this path at first

Reply:
I don't think we needed the lock in dispose